lib/auth/permissions.go

```diff
@@ -140,11 +144,22 @@ func (a *authorizer) authorizeRemoteUser(u teleport.RemoteUser) (*AuthContext, e
 
 // authorizeBuiltinRole authorizes builtin role
 func (a *authorizer) authorizeBuiltinRole(r teleport.BuiltinRole) (*AuthContext, error) {
-	return contextForBuiltinRole(r.Role)
+	// if recording will occur at proxy, the built-in role proxy will get
+	// additional permissions.
+	clusterConfig, err := a.config.GetClusterConfig()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var recordAtProxy bool
+	if clusterConfig.GetSessionRecording() == services.RecordAtProxy {
```

Comment:
I would assume two different builtin proxy roles based on whether proxy is recording or not, rather than modifying the role on the fly.